Development

notxcharles opened the pull request:
A player makes his way through a series of rooms. 
Each room can contain a monster and/or a weapon for the player. The player can pick up a weapon (which adds to their inventory) and equip it. The player must fight the monster in each room before they can 
progress to the next room. After completing 5 rooms, the player has completed and won the game.

notxcharles committed:
Room Dimensions

notxcharles committed:
Room Display

Can write the room to console, eg:
/ - - - - - \
| o o o o o |
| o o o o o |
| o o o o o |
\ - - - - - /

notxcharles committed:
Swapped x and y

notxcharles committed:
Update Room.cs

Increased each dimension by 2. This creates the initial intended behaviour of the room being of size x,y.
Previously, the "walls" of the room took up space so the room would display as smaller.
eg:
/ - - - - - \
| o o o o o |
| o o o o o |
| o o o o o |
| o o o o o |
| o o o o o |
| o o o o o |
| o o o o o |
\ - - - - - /

notxcharles committed:
Door, Treasure, Monster

Not Implemented Yet

notxcharles committed:
Assessment Checklist/TODO list

notxcharles committed:
DisplayRoom(), Better named variables

notxcharles committed:
RenderRoom() and DisplayRoom()

RenderRoom() writes the positions of the room in a 2d array, this.roomDisplay()

DisplayRoom() iterates over the 2d array and displays the room to the Console

notxcharles committed:
Changed roomSize

Changed how roomSize is calculated. Formally, the Room() class would +2 to each x and y dimension.
Now, the roomSize includes the positions where the walls would be.
roomSize now measures the outer perimeter rather than the inner perimeter

notxcharles committed:
doorIsLocked and isTreasure

Set to true or false depending on if there is a monster or treasure in the room

notxcharles committed:
Basic Game Plan

notxcharles committed:
Room Constructor Overloading

notxcharles committed:
Room coordinates origin is bottom left

Now, the room coordinates begin in the bottom left at (0,0)

notxcharles committed:
Pseudocode

Laying out logic that is needed

notxcharles committed:
GameStartDisplay()

notxcharles committed:
Add room names, welcome message

Added a room name to the room class.
Each room also welcomes the player to the room, displaying the room name and description

notxcharles committed:
Monster Class

Monster class creates an object of a monster. Each monster has a name, a breed, health and does damage if attacked.
GetAttackDamage() - the monster will do a random amount of damage, this is based on a normal distribution with the preset monster's attack damage as the mean.

notxcharles committed:
Getter for AverageAttackDamage

notxcharles committed:
Rooms with monsters must have Monster Class

Going forward, rooms with a monster must contain a monster class.
If a room contains treasure, it can't contain a monster or an exit door.

WelcomePlayer() welcomes the player to the room, displaying a custom message in the Console. This lets the player know the name of the room, description of the room and the monster that is in the room

notxcharles committed:
Player GetDecision()

GetDecision() prompts the player for their next action
0- display inventory
1- attack the monster in the room
2- open the door
9- exit the game
function will return 0,1,2 or exit the game.

notxcharles committed:
Player.ViewInventory()

Displays a message in the Console that shows a list of all contents in the player's inventory

notxcharles committed:
Player can progress to the next room

notxcharles committed:
Adjust Room's Welcome Message

notxcharles committed:
Rooms are created with random parameters

The name, description, dimensions, door position, monster and monster's position are randomised.

The monster randomly selects a name and a breed from a list.

notxcharles committed:
DisplayRoom()

DisplayRoom() displays without two blank lines. I was adding 2 to the size of the array leading to unnecessary memory usage

notxcharles committed:
Random Room name and description can

A random Room name and Room description can be automatically selected from a premade list

notxcharles committed:
Renaming of public class variables

notxcharles committed:
CreateMonsterRoom()

Creating rooms is now moved to a function CreateMonsterRoom().

notxcharles committed:
Class Weapon

The player may hold multiple Weapons in their inventory. Each individual weapon is given a type and a damage amount

notxcharles committed:
Weapon Summary()

notxcharles committed:
Create Weapon Class

notxcharles committed:
PickUpWeapon() and EquipDifferentWeapon()

PickUpWeapon() allows the player to pick up an additional weapon, unless they are already holding 4.

EquipDifferentWeapon() allows the player to change which weapon that is equipped with a weapon that is in their inventory

notxcharles committed:
Random Weapon Type

Similar to random room name or description, a new weapon can select a random type (from a valid list of weapons)

notxcharles committed:
Rooms now contain weapons

notxcharles committed:
Weapons in Rooms, More Decisions

Rooms are now created containing weapons.
The player has more decisions that they can make. They can choose to change their equipped weapon or pick up a new weapon

notxcharles committed:
Additional Options

notxcharles committed:
weaponNames renamed to weaponTypes

notxcharles committed:
Change equipped weapon

Player is given the option to change the weapon that their character has equipped. They have the choice of any weapon in their inventory. If they choose to change weapon, their current equipped weapon is added to their inventory and the new weapon is removed from inventory and added to the player's equipped weapon.

notxcharles committed:
Player can no longer pick up a weapon twice

notxcharles committed:
Player can opt to heal

notxcharles committed:
Fixed Gaussian random attack damage

notxcharles committed:
Player can fight the monsters

Defeating the monster will unlock the door in the room

notxcharles committed:
Logic and Grammar Errors

notxcharles committed:
Game can be completed

notxcharles committed:
Completed Tasks

notxcharles committed:
Remove Old Comments

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
 namespace DungeonExplorer
 {
     internal class Game
     {
```
good use of Encapsulation!

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+                    }
+                }
+
+                
```
a 5 second delay may make the game feel much slower than it has to be, perhaps consider reducing this

notxcharles replied:
Good suggestion, this is implemented now

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+            // When the player retrieves the loot they have won the game
+            FinishGame();
+            return;
+        }
```
The ansi escape sequence is great to improve the readability of your game .

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+            Console.ReadKey();
+            ClearConsole();
+            return;
+        }
```
This method is fairly long and I feel it can be broken up to improve readability

notxcharles replied:
I've removed DisplayRoom() which has resulted in this function being slimmed down. Display and RenderRoom() functions were no longer being used so this doesn t affect any functionality

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+                return new Room(roomName, roomDescription, roomDimensions, doorPosition, currentMonster, monsterCoords);
+            }
+            return new Room(roomDimensions, doorPosition, currentMonster, monsterCoords, weapon);
+        }
```
Great logic forcing the player to defeat monsters before progressing

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+            if (player.Health <= 0)
             {
-                // Code your playing logic here
+                Console.WriteLine($"The monster has killed you! You took {monsterAttackDamage} damage. Game Over");
```
Consider having a replay / restart option instead of an immediate end to the game.

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+                Console.WriteLine($"The monster has hit you for {monsterAttackDamage} damage. You now have {player.Health}/{player.MaxHealth}");
+            }
+            return;
+        }
```
i like this simple but effective way to wrap up the game!

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+            while (roomNumber < m_numberOfRooms)
+            {
+                m_currentRoom.WelcomePlayer(roomNumber);
+                int decision = m_player.GetDecision();
```
consider using switch statements here over excessive else ifs, in the future if more decisions are needed for implementation this could become fairly messy

Ashx500 commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+using System.Threading;
 
 namespace DungeonExplorer
 {
```
Good use of internal over public

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+namespace DungeonExplorer
+{
+    public class Monster
+    {
```
Good use of encapsulation

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+            "Cyclopes"
+        };
+
+
```
great overloaded constructor design

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+            MaxHealth = health;
+            AverageAttackDamage = averageAttack;
+            
+        }
```
private string GetMonsterName()
private string GetMonsterBreed()
private double CreateRandomGaussianNumber(int mean, int stdDev)
GetAttackDamage()

are all great and follow the single responsibility principle very well, i also like that the randomness, ensures good monster variety

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+        private string GetMonsterName()
+        {
+            int index = m_random.Next(0, m_monsterNames.Length);
```
i love the consideration for wanting to generate a more pseudorandom number but using this method is not good practice, try using a static random instance instead of re-initializing it in every object

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+        public int GetAttackDamage()
+        {
+            //m_averageAttack represents the mean of a normal distribution
+            //attackDamage will be a random datapoint in the distribution
```
int stdDevPercentage is a hard coded constant, consider using the following:

            private const int stdDevPercentage = 5;

Ashx500 commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+        public int MaxHealth { get; set; }
+        public int AverageAttackDamage { get; private set; }
+        private Random m_random;
+        private string[] m_monsterNames = new string[] {
```
Consider having Monster Names and Breeds in a .json file instead of having them hardcoded

notxcharles replied:
This adds some additional complexity which I am not willing to do at this stage, however I'll consider this for the next assessment

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
 namespace DungeonExplorer
 {
     public class Player
     {
```
Good use of encapsulation for inventory and currentEquippedWeapon

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+using System.Collections.Generic;
 
 namespace DungeonExplorer
 {
```
Good modular design for the player class making it easy to modify / extend

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+            //The player's default starting weapon are their fists
+            currentEquippedWeapon = new Weapon("Fists", 30);
+            m_random = new Random();
+        }
```
public void PickUpWeapon(Weapon weapon)
public void EquipDifferentWeapon(int weaponIndex)
public void ViewInventory()

are well organised and easy to read/understand

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
-        private List<string> inventory = new List<string>();
-
+        public int MaxHealth { get; private set; }
+        public int Health { get; set; }
```
Great use of a making a list for inventory, this is very flexible and scale-able

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
         {
             Name = name;
+            MaxHealth = health;
             Health = health;
```
consider making maxinventoryspace more dynamic by Letting the inventory size be configurable through the constructor or external config.

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+        private List<Weapon> inventory = new List<Weapon>();
+        public int maxInventorySpace { get; private set; }
+        private Weapon currentEquippedWeapon;
+        Random m_random;
```
You have both Health and MaxHealth, which is a bit redundant. consider using just MaxHealth for both, You can set Health equal to MaxHealth when healing

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+        {
+            return inventory.Count;
         }
-        public string InventoryContents()
```
GetAttackDamage() could be misleading since it returns the attack damage for the current weapon and not the player, while not a major issue consider renaming this to GetCurrentWeaponAttackDamage() to avoid confusion.

notxcharles replied:
Good catch!

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+            }
+            return -1;
+        }
+
```
good single responsibility method

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+            {
+                Console.WriteLine($"Items in your inventory:");
+                for (int i = 0; i < inventory.Count; i++)
+                {
```
inventory[i].CreateSummary, a very clean way to provide current weapon information

Ashx500 commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+                Console.WriteLine($"{weapon.Type} has been added to your inventory");
+            }
+            return;
+        }
```
another good single purpose method, very straightforward and easy to understand

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
 {
     public class Room
     {
-        private string description;
```
good use of encapsulation  for 

        private int[] roomDimensions;
        private int doorPosition;
        private int[] monsterCoords;
        private int[] treasureCoords;
        private char[,] roomDisplay;
        private bool isTreasure = true;
        private string[] m_roomNames = new string[]

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+            "A vast underground lake, the water impossibly still. Jagged rocks rise from the surface like teeth, and something beneath the water disturbs the reflection."
+        };
+        Random m_random;
+        // A monster and an exit door in this room
```
another amazing overloaded constructor, i have loved these throughout this project making it very flexible

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+            return;
+        }
+
+        //Renders the room and fills out the this.roomDisplay 2d array
```
The method for rendering the room is very impressive and adheres to SRP, really good work here Charles

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+        private int[] treasureCoords;
+        private char[,] roomDisplay;
+        private bool isTreasure = true;
+        private string[] m_roomNames = new string[] {
```
again i feel it would be good practice to have these different descriptions in a json file instead of being hardcoded

              "rooms": [
                  {
                    "name": "The Forgotten Hall",
                    "description": "A vast hall with towering stone pillars, their surfaces worn smooth by time. The air is thick with the scent of damp stone and something faintly metallic."
                  }

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+        }
+        private string GetRoomName()
+        {
+            int index = m_random.Next(0, m_roomNames.Length);
```
as previously mentioned i feel updating this and using a static random instance would be the way to go

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+            return Monster;
+        }
+
+        //Displays the contents of this.roomDisplay to the console
```
very impressive method here iterating through the room's 2D display array and printing it line by line, great logic for rendering a visual representation, however You should clarify what the dimensions represent and if the subtraction is needed, a comment explaining why would be helpful for the following:

       Console.WriteLine($"This is the room! A height of {xMax - 2} and a height of {yMax - 2}!");

Ashx500 commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+                }
+            }
+            return;
+        }
```
good oop practice for the welcomePlayer() method, however im slightly confused / curious of line 244 with //DisplayRoom(); commented out, if unneeded consider removing this however if you intend on using this and it serves a purpose its no issue

notxcharles replied:
Good catch. I no longer use DisplayRoom() as it is a bit unintuative to understand. I'll keep the functionality incase it is needed in the future but I'll remove the comment

Ashx500 commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+namespace DungeonExplorer
+{
+    public class Weapon
+    {
```
good encapsulation for  both
          
          public string Type { get; private set; }
          public int AverageAttackDamage { get; private set; }

Ashx500 commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+            "Car",
+            "Fireworks",
+            "Tennis Ball Machine"
+        };
```
good constructor overloading!

Ashx500 commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+            m_random = new Random();
+            Type = GetWeaponType();
+            AverageAttackDamage = weaponAverageDamage;
+        }
```
good use of private methods for internal logic for both of the following:

            private string GetWeaponType()
            private double CreateRandomGaussianNumber(int mean, int stdDev)

Ashx500 commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+            int index = m_random.Next(0, weaponTypes.Length);
+            Thread.Sleep(25); //Add a little Thread.Sleep() so that Random can be more pseudorandom
+            return weaponTypes[index];
+        }
```
i love the effort you have put in to make sure the damage values have realistic variation rather than just using a linear or uniformly distributed value.

Ashx500 commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+        public int GetAttackDamage()
+        {
+            //m_averageAttack represents the mean of a normal distribution
+            //attackDamage will be a random datapoint in the distribution
```
consider using 

          private const int StdDevPercentage = 5;

notxcharles replied:
Good idea

Ashx500 reviewed (approved):
overall a really impressive project you have here Charles, you have used Encapsulation very well, nearly all of your methods follow the single responsibility principle and your use of overloaded constructors is great for flexibility, I have left a few improvements some more important than others however all in all this is very impressive work, good job!

Mitchell-Nicholls commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+        private Room m_currentRoom;
+        private int m_numberOfRooms;
+        private Random m_random;
+        
```
Good use of encapsulation with private fields for necessary fields the user shouldn't be able to modify.

Mitchell-Nicholls commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+                }
+
+                
+                Thread.Sleep(5000);
```
I agree that this time could be considered of reducing as I believe from my experience of playing the game this made me feel like I was being interrupted after each option I chose.

notxcharles replied:
Implemented- the user now needs to press a key to advance to the next turn, rather than wait for 5 seconds

Mitchell-Nicholls commented on Game.cs:
```diff
@@ -1,26 +1,164 @@
+        {
+            Console.Clear();
+            Console.WriteLine("\x1b[3J");
+            return;
```
Great use of a method to reduce content on user's console which I appreciated.

Mitchell-Nicholls commented on Monster.cs:
```diff
@@ -0,0 +1,118 @@
+            "Wilson Fisk",
+            "Negan",
+            "Pete Campbell",
+            "Alaric Saltzman",
+            "Dexter Morgan",
+            "Tommen Baratheon",
+            "Annalise Keating",
+            "Jack Torrance",
+            "Jack Torrance",
+            "Omar Little",
+            "Bane",
+            "Regina Mills",
+            "Kai Parker",
+            "Dennis Reynolds",
+            "Darth Maul",
+            "Count Dooku",
+            "Jabba the Hutt",
+        };
+        private string[] m_monsterBreeds = new string[] {
+            "Goblin",
+            "Troll",
+            "Werewolf",
+            "Vampire",
+            "Zombie",
+            "Dragon",
+            "Skeleton",
+            "Ghoul",
+            "Banshee",
+            "Witch",
+            "Demon",
+            "Phantom",
+            "Hydra",
+            "Orc",
+            "Minotaur",
+            "Kraken",
+            "Giant",
+            "Hellhound",
+            "Ogre",
+            "Cyclopes"
+        };
```
I agree that the monsters and the breed of monsters stored in a file which was then imported would reduce the content overload of this file and make it easier to read in my opinion.

notxcharles replied:
This adds some additional complexity which I am not willing to do at this stage, however I'll consider this for the next assessment

Mitchell-Nicholls commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
-        public int Health { get; private set; }
-        private List<string> inventory = new List<string>();
-
+        public int MaxHealth { get; private set; }
+        public int Health { get; set; }
```
I appreciate the use of Max health and health to create a health bar to remind the user of the health they hold reminding them whether healing is the best choice to make. Additionally using encapsulation with getters and private setters.

notxcharles replied:
Thank you, I've also fixed the error showing the player's health to be 250/100, instead this is 250/250 (or whatever value MaxHealth is)

Mitchell-Nicholls commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+                            Console.WriteLine($"{key} was pressed. You must press a key that corresponds to a weapon");
+                        }
+                    }
+                    catch (FormatException e)
+                    {
+                        Console.WriteLine($"{key} was pressed. You may only press a key that corresponds to a weapon");
```
Good use of exception handling to identify when the user doesn't enter a valid input for a valid weapon index.

Mitchell-Nicholls commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+                    }
+                }
+                catch (FormatException e)
+                {
```
Good use of exception handling evident to be functional through testing preventing the user type anything they want preventing to create bugs with the problem.

Mitchell-Nicholls commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+            if (inventory.Count == 0)
+            {
+                Console.WriteLine("You have no items in your inventory. You can hold up to 4 weapons.");
+            }
```
I respect the inventory count constantly keeping track of the amount of items in the user's inventory preventing the user to hold more than this count. Very cleanly comprised and easy to understand the logic of how it works.

Mitchell-Nicholls commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+            "Hammer",
+            "Shotgun",
+            "Rifle",
+            "Bow and Arrow",
```
Again to take into consideration, have an imported file with the different weapon types in to randomise from.

notxcharles replied:
Thank you, I'll consider this for the future assignment

Mitchell-Nicholls commented on Room.cs:
```diff
@@ -1,17 +1,253 @@
+            "A spiral staircase leading downward, its steps uneven and slick with slime. The deeper one descends, the more the walls seem to close in.",
+            "A desolate chamber with cracked mirrors covering every surface, each reflecting something just slightly... off.",
+            "A frozen cavern where jagged icicles hang from the ceiling, the floor slick and treacherous. The air bites at exposed skin, and breath comes out in thick clouds.",
+            "A grand library, its shelves stretching impossibly high. Books lie scattered, some pages torn, others missing entirely. The silence is absolute.",
```
In my opinion again, to improve the layout and reduce the length of the program, importing a file containing the values in would help as you can then add new values into the file for more room descriptions.

Mitchell-Nicholls commented on Player.cs:
```diff
@@ -1,25 +1,163 @@
+                    }
+                    Console.WriteLine($"What weapon would you like to equip?");
+                    ConsoleKeyInfo key = Console.ReadKey();
+                    try
```
While I appreciate the option of changing the user's equipped weapon, I find that I was constantly using this option to compare the damages of the weapons with the constant repetition to check the weapon I was holding before I switched it. In my opinion, I believe if there was a way to check the weapon you're currently holding before switching it out would that this would be beneficial. Additionally, having an option for the user to reject switching out their weapon when they go on this option for the reason of using the switch equipped weapon to compare damages.

Mitchell-Nicholls commented on Weapon.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+            string summary = ($"{Type}, dealing an average of {AverageAttackDamage} per attack");
+            return summary;
+        }
```
I really appreciate the code being structured this way with different equations for the average attack damage the user's weapon does as well as having a randomiser of the weapon which is discoverable in the different rooms. Well done Charles. That is amazing to see.

Mitchell-Nicholls reviewed (approved):
In conclusion, I believe this project was heavily thought about and a lot of different factors were taken into account. Good use of getters and setters for different fields, simple but effective use of exception handling for different circumstances tested with effective use by me, amazing use of random rooms, room descriptions, weapon types and player and enemy damage. Additionally variations of methods for the user to pick up items and for the user to get a description of the current room is in place and functions as intended. 

However, I believe through previous comments from me I believe that certain things could be considered to be put in place including the option for the user to compare the weapons they currently have on them as I found that the damage was difficult to compare having to switch weapons to see the damage of the weapon I was holding initially and with the case of checking this damage, to provide an option for the user to choose not to switch their weapon for this reason of comparing damage. Minor thing I found but the user's health for me when testing displayed 250/100 when playing and whenever my health would go down would stay at blank/100.

Overall, this is an amazing project with a lot of thought from its author. Well done Charles.

notxcharles committed:
Max Line Length

To improve readability, long lines are now split

notxcharles committed:
private static Random

Using private static Random so that a new pseudorandom seed is generated each time a new random number is generated. this means that i no longer have to use thread.sleep() for a short period of time

notxcharles committed:
Health: {Health}/{MaxHealth}

Fixed bug that displayed MaxHealth as 100.

notxcharles committed:
Show current weapon damage

Now shows the currently equipped weapon's average damage after displaying the player's health.

notxcharles committed:
Max Inventory Size can be set through the constructor

notxcharles committed:
Changing weapons when inventory is empty no longer errors

Attempting to change equipped weapons when inventory is empty no longer causes an error.
Changed some names of formulas
Character details moved to it's own function, ShowCharacterDetails()

notxcharles committed:
Add .idea to .gitignore

notxcharles committed:
Rename GetAttackDamage() to GetCurrentAttackDamage()

notxcharles committed:
Remove commented out function call

notxcharles committed:
stdDevPercentage now const

notxcharles committed:
Update Weapon.cs

Remove some unneeded requirements.
GetWeaponType() to CreateWeaponType()

notxcharles committed:
Changes to function naming, seperated ShowWeaponsInInventory()

Seperated ShowWeaponsInInventory() into ViewItemsInInventory() and SelectWeaponInInventory() to reduce duplicated code

DHaigh2005 commented on Game.cs:
```diff
@@ -1,26 +1,168 @@
+                        m_currentRoom.WeaponPickedUp();
+                    }
+                }
+                else if (decision == 3)
```
Indent this into an if statement or add an &&, so if there is no monster in the room, attacking doesn't break the code.

notxcharles replied:
Thank you, the last commit solves this issue

DHaigh2005 commented on Player.cs:
```diff
@@ -1,25 +1,185 @@
+            Console.WriteLine("(0) View Inventory");
+            Console.WriteLine("(1) Change Equipped Item");
+            Console.WriteLine("(2) Pick up weapon");
+            Console.WriteLine($"(3) Attack Monster with {currentEquippedWeapon.Type}");
```
Indent this into an if statement so it doesn't give the user a prompt to attack when there is not a monster in the room

notxcharles replied:
All done! Thanks!

DHaigh2005 reviewed (changes requested):
You have used a consistent format when naming variables.
Your have used some comments which improve the readability of the code.
You have included everything requested in the brief such as classes, encapsulation and abstraction, methods and constructors, game interaction and also error checking and handling.

However, when no monster is present in the room, it still gives the user a prompt to attack the monster. When selected, the program stops flowing.
When the monsters health is on or below zero, it is still able to attack the player.

notxcharles committed:
Monster can only attack when alive

Monster can only attack and damage the player if the monster's health is positive

notxcharles committed:
User isn't prompted to attack when no monster

The user is no longer prompted to attack the monster if the monster isn't in the room or the monster is dead.

notxcharles committed:
Press key when turned finished

This replaces the previous logic to wait 5 seconds after each turn. Now, the user must press any key to advance. The user doesn't have to press any specific key

notxcharles committed:
All functions return

All functions, even void, now return to signifiy the end

eoin2406 commented on Game.cs:
```diff
@@ -1,26 +1,189 @@
-
+            m_gameName = gameName;
+            m_numberOfRooms = amountOfRooms;
+            m_player = player;
```
The variable names are quite inconsistent throughout all of the files. Whilst they do show their purposes, I think for clarity it would be better to make them all the same and more consistent.

notxcharles replied:
This has been changed in the latest commit, thanks

eoin2406 commented on Monster.cs:
```diff
@@ -0,0 +1,114 @@
+            "Tommen Baratheon",
+            "Annalise Keating",
+            "Jack Torrance",
+            "Jack Torrance",
```
The name "Jack Torrance" has 2 chances to be chosen, and whilst it is not too significant of a problem, it is something that would be better to be removed as it means all names have equal chance to be chosen.

eoin2406 commented on Monster.cs:
```diff
@@ -0,0 +1,114 @@
+        {
+            //m_averageAttack represents the mean of a normal distribution
+            //attackDamage will be a random datapoint in the distribution
+            int stdDevPercentage = 5;
```
I like the use of standard deviation allowing for different amounts of damage throughout the game, making it a bit more interesting.

eoin2406 commented on Player.cs:
```diff
@@ -1,25 +1,197 @@
+            //Check if the inventory is full (we'll say full=4 weapons)-
+            if (inventory.Count == maxInventorySpace)
+            {
+                Console.WriteLine("Your inventory is full! You cannot pick up any more weapons");
```
I like how it will tell you that you cannot add more items to the inventory once it has reached its maximum capacity.

eoin2406 commented on Player.cs:
```diff
@@ -1,25 +1,197 @@
+            Console.WriteLine("(1) Change Equipped Item");
+            Console.WriteLine("(2) Pick up weapon");
+            Console.WriteLine("(3) Retreat and heal");
+            Console.WriteLine("(4) Open the door");
```
Numbers being used for the options are great and makes it a lot easier and allows for significantly less user input.

eoin2406 commented on Player.cs:
```diff
@@ -1,25 +1,197 @@
+                    }
+                    else
+                    {
+                        Console.WriteLine($"{key} was pressed. You must press a key that " +
```
Good error handling here.

eoin2406 reviewed (changes requested):
The whole program is very intricate and has some very interesting methods that make the game feel much more complete and unique. The fact that the game will determine a random room, description, monster name and weapon is brilliant. You have great error-handling and the user input is minimal, making the game much more fast-paced and straightforward. Apart from some minimal issues, that could be easily resolved, it all is very consistent, clear and easy to read. Your error-handling is good, and does not bring up any errors when executed. I like how the room has the "GetDimensions" method, allowing for a more visual look on the room, making it feel much more interesting for the user.
Overall, the game seems very well-made, using neat methods that meet every requirement, as well as going above and beyond the expectations. The program follows the style guidelines correctly. I added a few comments throughout the code, hopefully some of the potential changes will help.

notxcharles committed:
Jack Torrance duplicate in list

notxcharles committed:
Class Member Variables have prefix m_

Class member variables have the prefix of m_ to help better distinguish them from the rest of the variables

notxcharles committed:
Remove Room.RenderRoom()

This is unneeded functionality, removing this will significantly reduce the length of the file and remove some clutter from view

notxcharles committed:
Private instance fields start with an underscore

Previously, class member variables had an m_ prefix before the variable name, now they just have an _ to adhere to c# formatting guidelines

notxcharles committed:
PascalCase for constant names

notxcharles committed:
Variable renaming, Comments some functions

notxcharles committed:
List of strings can be static

notxcharles committed:
Remove unused import statements

notxcharles committed:
Rename GetRoomDescription to CreateRoomDescription

This is also the case for some monster functions

notxcharles committed:
Option to view room description

aopacic25 commented:
Overall a very solid implementation which adheres to the brief requirements:

Classes and Objects:
- The program defines several classes: Game, Player, Room, Monster, and Weapon. ✓
- The Game class handles the game flow and initialises the player and one room. ✓
- The Player class tracks the player's name, health, and inventory. ✓
- The Room class represents several rooms, each with a unique description. ✓
- Objects are instantiated for Player, Room, Monster, and Weapon. ✓

Encapsulation and Abstraction:
- Private fields (_inventory in Player, RoomDescription in Room, for example) are encapsulated. ✓
- Public methods (GetAttackDamage, WelcomePlayer, for example) allow controlled access to data. ✓

Methods and Constructors:
- Constructors initialize objects (Player, Room, Monster, for example). ✓
- Methods like PickUpWeapon, GetAttackDamage, and WelcomePlayer are implemented. ✓

Basic Game Interaction:
- The player can:
   - Explore rooms. ✓
   - Check their inventory, health. ✓

Error Checking:
- The GetTurnDecisions method handles invalid input. ✓

I like the fact that that Weapon and Monster classes both create unique instances via randomisation, which translates well into more dynamic combat by giving them varying attack damage. Combat logic is also well-implemented through the PlayerFightsMonsters method. Overall interactivity is good, since the player can both pick up and change equipped weapons.

notxcharles committed:
Debug.Assert to test assumptions are true

notxcharles committed:
Debug.Asset and Testing Class

notxcharles committed:
Testing Class

notxcharles committed:
Fix logic errors

notxcharles committed:
Debug.Assert and Logic Errors

notxcharles committed:
C# Documentation strings

notxcharles committed:
Monster.cs docustrings

notxcharles committed:
Monster class docustring

notxcharles committed:
Weapon.cs docustrings

notxcharles committed:
Room.cs Documentation Comments

notxcharles committed:
Remove un required code from testing class

notxcharles committed:
Rename to PickUpItem

notxcharles committed:
Remove old commented code

notxcharles committed:
Update Player.cs
